// @camunda8/orchestration-cluster-api/dist/logger.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/logger.ts
var logger_exports = {};
__export(logger_exports, {
  createLogger: () => createLogger
});
module.exports = __toCommonJS(logger_exports);

// src/runtime/logger.ts
var ORDER = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
  debug: 4,
  trace: 5,
  silly: 6
};
function createLogger(opts = {}) {
  let currentLevel = opts.level || "error";
  let transport = opts.transport;
  const baseScope = opts.scope || "";
  function isEnabled(need) {
    return ORDER[currentLevel] >= ORDER[need];
  }
  function evalArgs(args) {
    return args.map((a) => typeof a === "function" && a.length === 0 ? a() : a).flat();
  }
  function emit(level, scope, rawArgs) {
    if (!isEnabled(level)) return;
    const args = evalArgs(rawArgs);
    const evt = { level, scope, ts: Date.now(), args };
    if (transport) {
      try {
        transport(evt);
      } catch {
      }
    } else {
      const tag = `[camunda-sdk][${level}]${scope ? `[${scope}]` : ""}`;
      const method = level === "error" ? "error" : level === "warn" ? "warn" : "log";
      console[method](tag, ...args);
    }
  }
  function emitCode(level, scope, code, msg, data) {
    if (!isEnabled(level)) return;
    const evt = { level, scope, ts: Date.now(), args: [msg], code, data };
    if (transport) {
      try {
        transport(evt);
      } catch {
      }
    } else {
      const tag = `[camunda-sdk][${level}]${scope ? `[${scope}]` : ""}`;
      const method = level === "error" ? "error" : level === "warn" ? "warn" : "log";
      console[method](tag, code + ":", msg, data ?? "");
    }
  }
  const make = (scope) => ({
    level: () => currentLevel,
    setLevel(l) {
      currentLevel = l;
    },
    setTransport(t) {
      transport = t;
    },
    error: (...a) => emit("error", scope, a),
    warn: (...a) => emit("warn", scope, a),
    info: (...a) => emit("info", scope, a),
    debug: (...a) => emit("debug", scope, a),
    trace: (...a) => emit("trace", scope, a),
    silly: (...a) => emit("silly", scope, a),
    scope(child) {
      return make(scope ? `${scope}:${child}` : child);
    },
    code(l, code, msg, data) {
      emitCode(l, scope, code, msg, data);
    }
  });
  return make(baseScope);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createLogger
});
