// @camunda8/sdk/dist/oca/lifters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrchestrationLifters = void 0;
/**
 * Orchestration Cluster API lifter namespace.
 *
 * This groups all branded key lifter namespaces from the Orchestration Cluster API
 * behind a single export so they don't pollute the root SDK namespace. Each lifter
 * exposes helper functions (e.g. `assumeExists(value: string)`) to convert legacy
 * string identifiers into their strongly-typed branded equivalents.
 *
 * Example:
 * ```typescript
 * import { OrchestrationLifters } from '@camunda8/sdk'
 * const processInstanceKey = OrchestrationLifters.ProcessInstanceKey.assumeExists('2251799813685249')
 * // processInstanceKey is now a branded ProcessInstanceKey type
 * ```
 */
/* eslint-disable @typescript-eslint/no-namespace */
const orchestration_cluster_api_1 = require("@camunda8/orchestration-cluster-api");
exports.OrchestrationLifters = {
    AuthorizationKey: orchestration_cluster_api_1.AuthorizationKey,
    BatchOperationKey: orchestration_cluster_api_1.BatchOperationKey,
    DecisionDefinitionId: orchestration_cluster_api_1.DecisionDefinitionId,
    DecisionDefinitionKey: orchestration_cluster_api_1.DecisionDefinitionKey,
    DecisionEvaluationInstanceKey: orchestration_cluster_api_1.DecisionEvaluationInstanceKey,
    DecisionEvaluationKey: orchestration_cluster_api_1.DecisionEvaluationKey,
    DecisionInstanceKey: orchestration_cluster_api_1.DecisionInstanceKey,
    DecisionRequirementsKey: orchestration_cluster_api_1.DecisionRequirementsKey,
    DeploymentKey: orchestration_cluster_api_1.DeploymentKey,
    DocumentId: orchestration_cluster_api_1.DocumentId,
    ElementId: orchestration_cluster_api_1.ElementId,
    ElementInstanceKey: orchestration_cluster_api_1.ElementInstanceKey,
    EndCursor: orchestration_cluster_api_1.EndCursor,
    FormId: orchestration_cluster_api_1.FormId,
    FormKey: orchestration_cluster_api_1.FormKey,
    IncidentKey: orchestration_cluster_api_1.IncidentKey,
    JobKey: orchestration_cluster_api_1.JobKey,
    MessageCorrelationKey: orchestration_cluster_api_1.MessageCorrelationKey,
    MessageKey: orchestration_cluster_api_1.MessageKey,
    MessageSubscriptionKey: orchestration_cluster_api_1.MessageSubscriptionKey,
    ProcessDefinitionId: orchestration_cluster_api_1.ProcessDefinitionId,
    ProcessDefinitionKey: orchestration_cluster_api_1.ProcessDefinitionKey,
    ProcessInstanceKey: orchestration_cluster_api_1.ProcessInstanceKey,
    ScopeKey: orchestration_cluster_api_1.ScopeKey,
    SignalKey: orchestration_cluster_api_1.SignalKey,
    StartCursor: orchestration_cluster_api_1.StartCursor,
    Tag: orchestration_cluster_api_1.Tag,
    TenantId: orchestration_cluster_api_1.TenantId,
    Username: orchestration_cluster_api_1.Username,
    UserTaskKey: orchestration_cluster_api_1.UserTaskKey,
    VariableKey: orchestration_cluster_api_1.VariableKey,
};
